File: solutions/clojure/078-clone-graph.clj
;; Problem 78: Clone Graph (LeetCode 133)
;; Difficulty: Med
;; Language: Clojure
;; 
(defn clone-graph [node]
  (letfn [(dfs [n visited]
            (if (contains? visited (:val n))
              [(visited (:val n)) visited]
              (let [clone {:val (:val n) :neighbors []}
                    vis (assoc visited (:val n) clone)
                    [nbrs vis2] (reduce (fn [[acc vs] nb]
                                          (let [[cn vs2] (dfs nb vs)]
                                            [(conj acc cn) vs2]))
                                        [[] vis] (:neighbors n))]
                [(assoc clone :neighbors nbrs) vis2])))]
    (first (dfs node {}))))

File: solutions/ocaml/078-clone-graph.ml
(*
  Problem 78: Clone Graph (LeetCode 133)
  Difficulty: Med
  Language: OCaml
*)
let clone_graph node =
  let visited = Hashtbl.create 16 in
  let rec dfs n =
    match Hashtbl.find_opt visited n.v with
    | Some c -> c
    | None ->
      let clone = { v = n.v; neighbors = [] } in
      Hashtbl.add visited n.v clone;
      clone.neighbors <- List.map dfs n.neighbors;
      clone
  in dfs node
